### ISSUE TITLE
Switch source exports to background process and cache

### ISSUE THREAD

**Issue by paynejd** [OpenConceptLab/oclapi#105]
It is not currently possible to request a full export of a source or to fetch other large datasets due to the extended processing time. Even if the processing time were reduced, handling very long-running processes as web requests is not a great approach. This ticket focuses on fetching full exports of sources only -- fetching other large datasets is out of scope for this phase.

Rough Draft Export API Specs here: https://github.com/OpenConceptLab/oclapi/wiki/Export-API

### COMBINING TEXT FROM EMAILS HERE --- WILL CHANGE

Preparation of an export will fire up a celery process.

The background process would adjust the API to support:
* request export
* get export status
* download latest export

<hr>

A last updated date is included in the response header of files, which can be used to fetch only the diff between the export and the latest version of the source or collection.

How is a diff represented?  Specifically, how do I differentiate between a create vs. an update vs. a delete?

If new mappings are added to the latest version of a source and then an export is recreated, will the recreated export be different than the original one? i.e. include the new mappings?
How are mappings stored in source versions?
I would think so.  Mappings are represented within a source version as a list of Object IDs.

What is the correct format/filetype field to include in the request/response headers?
I've been using 'application/zip'.

What is the best way to handle a request that has too many results that its going to break? How do we fail gracefully?
Is this still a concern?  Do you mean if the attachment is too large to transmit over HTTP?

How do we allow an administrator to still export a diff if its too large to export via an HTTP request? e.g. can we still trigger this to be processed in the celery environment? Can we do this via the API? Creating diffs between versions is important and not currently supported.
I believe the new design addresses this.

How do we get the latest version info for a source only? e.g. GET /orgs/CIEL/sources/CIEL/versions/?limit=1 Is there a way to sort this?
/orgs/CIEL/sources/CIEL/latest/

Supported HTTP Verbs:
POST - triggers OCL to create / recreate export file
GET - returns the file if its ready; triggers creation of file if it doesn't exist; returns different status/error codes based on the result
HEAD - returns whether file exists?
DELETE - deletes the export file
I'm not sure this is the best approach.  I think the POST should return a token unique to that export, and GET and HEAD should accept the token.  This way, there is no ambiguity between a POST and subsequent GET/HEAD calls.

Request header:
Authorization: Token ...
Compress: true
Format/filetype - ??
The "Compress" header shouldn't be necessary, right?  We'll always compress?

Parameters - NONE? Should "includeRetired" be supported?
I would think all the same parameters as a standard GET call.

**Comment by paynejd** [comment]
- Only support full exports of source versions, not the latest/head of a source.

Disallow this, even if the user specifies the latest source version id or the keyword "latest"?
…
 
- When a source version is "published" a cached export will be created

Sounds reasonable.  Have you given any thought to where you want to store these?  Might I suggest S3?
…
 
- The export will need to include the correct "last_updated" value

Not sure what this means, but I was thinking we'd embed the time of creation in the export file name.
…
 
- Subsequent changes could be retrieved from the latest version of the source using the old export methodology

Not sure what this means either.

**Comment by paynejd** [comment]
As a consumer of OCL I would prefer to just request a dictionary/collection/etc, and get it. I don't want to have to go through some request+polling process.

As an architect, I'd suggest optimizing so that 90% of queries are for things that are fast/inexpensive/cached. And avoid going down a path where lots of users are making large ad-hoc queries.

Definitely there will be >1 dictionary to be served, and it will be important to support downloading collections/subsets. (E.g. PIH is going to want to have a PIH dictionary, but then collections like "Mirebalais concepts", "Basic Haiti concepts", etc that pull concepts from PIH and CIEL.) I don't think that Rafal's "periodic build" suggestion is right, but a dictionary curator typically makes an explicit decision to publish versions at particular times. And we should push people to be mainly downloading those published versions, which can be pre-built and cached, so those requests are served statically. (I'd rather GET "/PIH/collection/basic-haiti-concepts/version/17" rather than be doing this as an ad-hoc query that I have to poll for being ready.)

Business-wise I could imagine that you only can have a certain number of downloadable versions in the free service.

**Comment by paynejd** [comment]
I think I have a better idea for the export GET response; in fact, it would obviate the need for a separate HEAD implementation.

S3 allows you to generate a temporary URL for any file you have access to.  So, in the header (for both GET and HEAD), we could simply return the temporary URL, or null, if the file hasn't been generated.  (I arbitrarily chose 60 seconds as the TTL for the URL.)

If that's okay with you, I'm ready to declare my work done.  I put in 8 hours between last night and today, in addition to the 2 hours I worked Thursday night.  I'm only telling you this, because I did not have a chance to implement the new requirements we discussed on the call yesterday, namely:
Results from Discussion

Maximum limit=500 for all GET requests - error if limit is set to an amount above that
Paging needs to be supported for GET source details with includeMappings and includeConcepts set to true
Source versions
Add support for "released" URL parameter to source versions GET request (e.g. GET /orgs/CIEL/sources/CIEL/versions/?released=true)
Add default sort of created_on descending for source versions
Goal is that a subscription client could get the most recent published source version using the following: GET /orgs/CIEL/sources/CIEL/versions/?limit=1&released=true
